### PR TITLE
Velero: restic plugin add tolerations

### PR DESCRIPTION
**Pull request by knrt10**
This commit adds a new attribute `tolerations` which will add tolerations to restic DaemonSet.

Since the cluster has multiple node pools with different purposes, it
adds an ability to deploy the restic DaemonSet on the nodes that need to
backup the PVC.

closes: #1334
Signed-off-by: knrt10 <kautilya@kinvolk.io>

### CODE REVIEW

**Review by surajssd — CHANGES_REQUESTED** [review]
Add a config conversion test.

**Comment by knrt10** [comment]
@surajssd added

**Comment by invidian** [comment]
>@surajssd added

I don't see tests being added.

**Comment by knrt10** [comment]
Just to confirm, by config conversion doesn't it mean to add tolerations to restic config test and check whether it is right or not?

**Comment by invidian** [comment]
>Just to confirm, by config conversion doesn't it mean to add tolerations to restic config tests and check whether it is right or not?

I think what @surajssd meant was to add tests, that toleration actually appears in the right place in the rendered manifests.

**Comment by knrt10** [comment]
@surajssd added the tests. Please check

File: pkg/components/velero/component_test.go [hunk 1]
```diff
@@ -195,3 +196,49 @@ component "velero" {
...
+      bucket   = "foo"
+      provider = "aws"
+    }
+		
+    tolerations {
+      key = "TestResticToletrationKey"
+      value = "TestResticToletrationValue"
+      operator = "Equal"
+      effect = "NoSchedule"
+      toleration_seconds = "1"
```
Comment (invidian) [inline pkg/components/velero/component_test.go:217]: Indentation.

File: pkg/components/velero/component_test.go [hunk 2]
```diff
@@ -195,3 +196,49 @@ component "velero" {
...
+
+	m, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	renderedDaemonSet := m["velero/templates/restic-daemonset.yaml"]
+	if !strings.Contains(renderedDaemonSet, "key: TestResticToletrationKey") {
+		t.Fatalf("Rendered manifests should have toleration set.")
+	}
```
Comment (invidian) [inline pkg/components/velero/component_test.go:243]: I suggest we use https://github.com/kinvolk/lokomotive/blob/efe364ea57e4d70e063ef6dbbeb304dc5a94f717/pkg/components/contour/component_test.go#L123-L126, which should be more accurate.

File: pkg/components/velero/component_test.go [hunk 3]
```diff
@@ -195,3 +196,49 @@ component "velero" {
...
+
+	d = component.LoadConfig(body, &hcl.EvalContext{})
+	if d.HasErrors() {
+		t.Fatalf("Valid config should not return error, got: %s", d)
+	}
+
+	m, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
```
Comment (invidian) [inline pkg/components/velero/component_test.go:237]: Hm, we usually use `%v` for error formatting in tests.

File: pkg/components/velero/component_test.go [hunk 4]
```diff
@@ -195,3 +196,49 @@ component "velero" {
...
+      toleration_seconds = "1"
+    }
+  }
+}
+`
+
+	component := NewConfig()
+
+	body, d := util.GetComponentBody(configHCL, Name)
+	if d != nil {
```
Comment (invidian) [inline pkg/components/velero/component_test.go:226]: `d` is type `hcl.Diagnostics`, we should use `HasErrors()` instead of `!=  nil`.

**Review by invidian — CHANGES_REQUESTED** [review]

File: pkg/components/velero/component_test.go [hunk 5]
```diff
@@ -195,3 +196,47 @@ component "velero" {
...
+
+	component := NewConfig()
+
+	body, d := util.GetComponentBody(configHCL, Name)
+	if d.HasErrors() {
+		t.Fatalf("Error getting component body: %v", d)
+	}
+
+	d = component.LoadConfig(body, &hcl.EvalContext{})
+	if d.HasErrors() {
```
Comment (invidian) [inline pkg/components/velero/component_test.go:231]: We can put it in single line.

**Review by invidian — APPROVED** [review]
Just last nit, otherwise LGTM

**Review by invidian — APPROVED** [review]

**Review by surajssd — APPROVED** [review]
LGTM Thanks.

**Pull request merged**